tests/sentinel/tests/helpers/check_leaked_fds.tcl

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env tclsh
+#
+# This script detects file descriptors that have leaked from a parent process.
+#
+# Our goal is to detect file descriptors that were opened by the parent and
+# not cleaned up prior to exec(), but not file descriptors that were inherited
+# from the grandparent which the parent knows nothing about. To do that, we
+# look up every potential leak and try to match it against open files by the
+# grandparent process.
+
+# Get PID of parent process
+proc get_parent_pid {proc} {
+    set fd [open "/proc/$proc/status" "r"]
+    set content [read $fd]
+    close $fd
+
+    if {[regexp {\nPPid:\s+(\d+)} $content _ ppid]} {
+        return $ppid
+    }
+
+    error "failed to get parent pid"
+}
+
+# Look up fdlink in the specified process's open file list
+proc proc_has_fdlink {proc fdlink} {
+    foreach fd [glob -tails -directory "/proc/$proc/fd" *] {
+        if { [catch {set s [file readlink "/proc/$proc/fd/$fd"]} err] } {
+            continue
+        }
+        if {$s == $fdlink} {
```

Comment:
if both the grandparent and us opened the same file (e.g. `/dev/urandom`) we'll miss it.
i think we should match the FD number too (not just where it links to)

Reply:
Good idea, change